lib/action_view/component/base.rb

```diff
@@ -110,40 +110,41 @@ def call_method_name(variant)
           end
         end
 
-        def has_initializer?
-          self.instance_method(:initialize).owner == self
-        end
-
         def source_location
-          # Require #initialize to be defined so that we can use
-          # method#source_location to look up the file name
-          # of the component.
-          #
-          # If we were able to only support Ruby 2.7+,
-          # We could just use Module#const_source_location,
-          # rendering this unnecessary.
-          raise NotImplementedError.new("#{self} must implement #initialize.") unless has_initializer?
-
-          instance_method(:initialize).source_location[0]
+          @source_location ||=
+              begin
+                # Require #initialize to be defined so that we can use
+                # method#source_location to look up the file name
+                # of the component.
+                #
+                # If we were able to only support Ruby 2.7+,
+                # We could just use Module#const_source_location,
+                # rendering this unnecessary.
+                #
+                initialize_method = instance_method(:initialize)
+                initialize_method.source_location[0] if initialize_method.owner == self
+              end
         end
 
         def eager_load!
-          self.descendants.each do |descendant|
-            descendant.compile if descendant.has_initializer?
-          end
+          self.descendants.each(&:compile)
         end
 
         def compiled?
           @compiled && ActionView::Base.cache_template_loading
         end
 
+        def compile!
+          compile(true)
+        end
+
         # Compile templates to instance methods, assuming they haven't been compiled already.
         # We could in theory do this on app boot, at least in production environments.
         # Right now this just compiles the first time the component is rendered.
-        def compile
+        def compile(raise_if_invalid = false)
```

Comment:
I suggest following the same protocol of `save!` method and use `validate: false` as argument.

Reply:
Agreed. @bostanio mind making this change?